### PR TITLE
Fixed Member Caching, Message#member, Member#roles

### PR DESCRIPTION
**Pull request by DjDeveloperr**
In this PR I have fixed Member Caching and Message#member, Member#roles properties, and added some JS Doc comments.
And also reworked REST Manager with better errors and handling all over.

### CODE REVIEW

File: src/gateway/handlers/messageCreate.ts [hunk 1]
```diff
@@ -17,11 +17,26 @@ export const messageCreate: GatewayEventHandler = async (
...
+    d.member.user = d.author
+    await guild.members.set(d.author.id, d.member)
+    member = await guild.members.get(d.author.id)
+  }
   const mentions = new MessageMentions()
   const message = new Message(gateway.client, d, channel as any, user, mentions)
+  message.member = member
   if (guild !== undefined) message.guild = guild
+  if (message.member !== undefined) {
+    if(message.member.user === undefined) {
```
Comment (AkiaCode) [inline src/gateway/handlers/messageCreate.ts:34]: if( ......

File: src/gateway/handlers/messageCreate.ts [hunk 2]
```diff
@@ -17,11 +17,26 @@ export const messageCreate: GatewayEventHandler = async (
...
+      const user = await gateway.client.users.get(message.member.id)
+      if(user !== undefined) {
```
Comment (AkiaCode) [inline src/gateway/handlers/messageCreate.ts:36]: if( ......

File: src/managers/members.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import { User } from "../structures/user.ts";
```
Comment (AkiaCode) [inline src/managers/members.ts:1]: Change " " to ' '

File: src/managers/members.ts [hunk 4]
```diff
@@ -13,11 +14,29 @@ export class MembersManager extends BaseManager<MemberPayload, Member> {
...
   }
 
+  async get (key: string): Promise<Member | undefined> {
+    const raw = await this._get(key)
+    if (raw === undefined) return
+    const user = new User(this.client, raw.user)
+    const res = new this.DataType(this.client, raw, user)
+    for (const roleid of res.roleIDs as string[]) {
+      const role = await this.guild.roles.get(roleid)
+      if(role !== undefined) res.roles.push(role)
```
Comment (AkiaCode) [inline src/managers/members.ts:24]: if( ....

File: src/managers/members.ts [hunk 5]
```diff
@@ -13,11 +14,29 @@ export class MembersManager extends BaseManager<MemberPayload, Member> {
...
-      this.client.rest.get(GUILD_MEMBER(this.guild.id, id)).then(data => {
-        this.set(id, data as MemberPayload)
-        resolve(new Member(this.client, data as MemberPayload))
+      this.client.rest.get(GUILD_MEMBER(this.guild.id, id)).then(async data => {
+        await this.set(id, data as MemberPayload)
+        const user: User = new User(this.client, data.user)
+        const res = new Member(this.client, data as MemberPayload, user)
+        for (const roleid of res.roleIDs as string[]) {
+          const role = await this.guild.roles.get(roleid)
+          if(role !== undefined) res.roles.push(role)
```
Comment (AkiaCode) [inline src/managers/members.ts:37]: if( ....

File: src/structures/member.ts [hunk 6]
```diff
@@ -1,26 +1,28 @@
-import cache from '../models/cache.ts'
 import { Client } from '../models/client.ts'
 import { MemberPayload } from '../types/guild.ts'
 import { Base } from './base.ts'
+import { Role } from "./role.ts"
```
Comment (AkiaCode) [inline src/structures/member.ts:4]: Change " " to ' '

File: src/test/cmd.ts [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import { CommandClient, Intents } from '../../mod.ts';
 import PingCommand from "./cmds/ping.ts";
+import UserinfoCommand from "./cmds/userinfo.ts";
```
Comment (AkiaCode) [inline src/test/cmd.ts:3]: - Change UserinforCommand to UserInfoCommand
- Remove `;`
- Change " " to ' '

File: src/test/cmd.ts [hunk 8]
```diff
@@ -13,6 +14,9 @@ client.on('ready', () => {
   console.log(`[Login] Logged in as ${client.user?.tag}!`)
 })
 
+client.on("commandError", console.log)
```
Comment (AkiaCode) [inline src/test/cmd.ts:17]: Change " " to ' '

File: src/test/cmds/userinfo.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { Command, Member, CommandContext, Embed } from "../../../mod.ts";
```
Comment (AkiaCode) [inline src/test/cmds/userinfo.ts:1]: - Change " " to ' '
- Remove `;`

File: src/test/cmds/userinfo.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+export default class UserinfoCommand extends Command {
+    name = "userinfo"
+    guildOnly = true
+
+    execute(ctx: CommandContext): void {
+        const member: Member = ctx.message.member as any
+        const embed = new Embed()
+        .setTitle(`User Info`)
+        .setAuthor({ name: member.user.tag })
+        .addField("ID", member.id)
```
Comment (AkiaCode) [inline src/test/cmds/userinfo.ts:12]: Change " " to ' '

File: src/test/cmds/userinfo.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+        .addField("Roles", member.roles.map(r => r.name).join(", "))
```
Comment (AkiaCode) [inline src/test/cmds/userinfo.ts:13]: Change " " to ' '

**Review by AkiaCode — CHANGES_REQUESTED** [review]

File: src/managers/emojis.ts [hunk 12]
```diff
@@ -10,6 +10,17 @@ export class EmojisManager extends BaseManager<EmojiPayload, Emoji> {
     super(client, `emojis`, Emoji)
   }
 
+  async get (key: string): Promise<Emoji | undefined> {
+    const raw = await this._get(key)
+    if (raw === undefined) return
+    const emoji = new this.DataType(this.client, raw)
+    if((raw as any).guild_id !== undefined) {
```
Comment (AkiaCode) [inline src/managers/emojis.ts:17]: if(

File: src/managers/emojis.ts [hunk 13]
```diff
@@ -10,6 +10,17 @@ export class EmojisManager extends BaseManager<EmojiPayload, Emoji> {
...
+      const guild = await this.client.guilds.get((raw as any).guild_id)
+      if(guild !== undefined) emoji.guild = guild
```
Comment (AkiaCode) [inline src/managers/emojis.ts:19]: if(

File: src/managers/guildEmojis.ts [hunk 14]
```diff
@@ -0,0 +1,91 @@
+import { Client } from '../models/client.ts'
+import { Emoji } from "../structures/emoji.ts"
```
Comment (AkiaCode) [inline src/managers/guildEmojis.ts:2]: " "

File: src/managers/guildEmojis.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+import { Guild } from '../structures/guild.ts'
+import { Role } from "../structures/role.ts"
+import { EmojiPayload } from "../types/emoji.ts"
```
Comment (AkiaCode) [inline src/managers/guildEmojis.ts:5]: " "

File: src/managers/guildEmojis.ts [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+import { CHANNEL, GUILD_EMOJI, GUILD_EMOJIS } from '../types/endpoint.ts'
+import { BaseChildManager } from './baseChild.ts'
+import { EmojisManager } from "./emojis.ts"
```
Comment (AkiaCode) [inline src/managers/guildEmojis.ts:8]: " "

File: src/managers/guildEmojis.ts [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    this.guild = guild
+  }
+
+  async get(id: string): Promise<Emoji | undefined> {
+    const res = await this.parent.get(id)
+    if (res !== undefined && res.guild?.id === this.guild.id) return res
+    else return undefined
+  }
+
+  async delete(id: string): Promise<boolean> {
```
Comment (AkiaCode) [inline src/managers/guildEmojis.ts:28]: delete (

File: src/managers/guildEmojis.ts [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+    return this.client.rest.delete(CHANNEL(id))
+  }
+
+  async fetch(id: string): Promise<Emoji | undefined> {
```
Comment (AkiaCode) [inline src/managers/guildEmojis.ts:32]: fetch (

File: src/managers/guildEmojis.ts [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+          data.guild_id = this.guild.id
+          await this.set(id, data as EmojiPayload)
+          emoji.guild = this.guild
+          resolve(emoji)
+        })
+        .catch(e => reject(e))
+    })
+  }
+
+  async create(name: string, url: string, roles?: Role[] | string[] | string): Promise<Emoji | undefined> {
```
Comment (AkiaCode) [inline src/managers/guildEmojis.ts:47]: create (

File: src/managers/guildEmojis.ts [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+          data.guild_id = this.guild.id
+          await this.set(data.id, data as EmojiPayload)
+          emoji.guild = this.guild
+          resolve(emoji)
+        })
+        .catch(e => reject(e))
+    })
+  }
+
+  async array(): Promise<Emoji[]> {
```
Comment (AkiaCode) [inline src/managers/guildEmojis.ts:77]: array (

File: src/managers/guildEmojis.ts [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    const arr = (await this.parent.array()) as Emoji[]
+    return arr.filter(
+      (c: any) => c.guild !== undefined && c.guild.id === this.guild.id
+    ) as any
+  }
+
+  async flush(): Promise<boolean> {
```
Comment (AkiaCode) [inline src/managers/guildEmojis.ts:84]: flush (

File: src/structures/emoji.ts [hunk 22]
```diff
@@ -2,13 +2,15 @@ import { Client } from '../models/client.ts'
 import { EmojiPayload } from '../types/emoji.ts'
 import { USER } from '../types/endpoint.ts'
 import { Base } from './base.ts'
+import { Guild } from "./guild.ts"
```
Comment (AkiaCode) [inline src/structures/emoji.ts:5]: " "

File: src/structures/emoji.ts [hunk 23]
```diff
@@ -20,17 +22,16 @@ export class Emoji extends Base {
     } else return `<a:${this.name}:${this.id}>`
   }
 
+  toString(): string {
```
Comment (AkiaCode) [inline src/structures/emoji.ts:25]: toString ()

File: src/structures/emoji.ts [hunk 24]
```diff
@@ -20,17 +22,16 @@ export class Emoji extends Base {
...
+    return this.getEmojiString
+  }
+
   constructor (client: Client, data: EmojiPayload) {
     super(client, data)
     this.id = data.id
     this.name = data.name
+    if(data.user !== undefined) this.user = new User(this.client, data.user)
```
Comment (AkiaCode) [inline src/structures/emoji.ts:33]: if (

File: src/structures/guild.ts [hunk 25]
```diff
@@ -6,7 +6,8 @@ import { VoiceState } from './voiceState.ts'
 import { RolesManager } from '../managers/roles.ts'
 import { GuildChannelsManager } from '../managers/guildChannels.ts'
 import { MembersManager } from '../managers/members.ts'
-import { Emoji } from "./emoji.ts"
+import { Role } from "./role.ts"
```
Comment (AkiaCode) [inline src/structures/guild.ts:9]: " "

File: src/structures/guild.ts [hunk 26]
```diff
@@ -6,7 +6,8 @@ import { VoiceState } from './voiceState.ts'
...
+import { GuildEmojisManager } from "../managers/guildEmojis.ts"
```
Comment (AkiaCode) [inline src/structures/guild.ts:10]: " "

File: src/structures/guild.ts [hunk 27]
```diff
@@ -208,4 +210,8 @@ export class Guild extends Base {
         data.approximate_presence_count ?? this.approximatePresenceCount
     }
   }
+
+  async getEveryoneRole(): Promise<Role> {
```
Comment (AkiaCode) [inline src/structures/guild.ts:214]: getEveryoneRole ()

File: src/test/cmds/addemoji.ts [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+  name = 'addemoji'
+  aliases = [ 'ae', 'emojiadd' ]
+  args = 2
+  guildOnly = true
+
+  execute(ctx: CommandContext): any {
+    const name = ctx.args[0]
+    if(name === undefined) return ctx.message.reply('No name was given!')
+    const url = ctx.argString.slice(name.length).trim()
+    if(url === '') return ctx.message.reply('No URL was given!')
```
Comment (AkiaCode) [inline src/test/cmds/addemoji.ts:14]: if (

File: src/test/cmds/addemoji.ts [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+
+export default class AddEmojiCommand extends Command {
+  name = 'addemoji'
+  aliases = [ 'ae', 'emojiadd' ]
+  args = 2
+  guildOnly = true
+
+  execute(ctx: CommandContext): any {
+    const name = ctx.args[0]
+    if(name === undefined) return ctx.message.reply('No name was given!')
```
Comment (AkiaCode) [inline src/test/cmds/addemoji.ts:12]: if (

File: src/test/cmds/addemoji.ts [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+    const url = ctx.argString.slice(name.length).trim()
+    if(url === '') return ctx.message.reply('No URL was given!')
+    ctx.message.guild?.emojis.create(name, url).then(emoji => {
+        if(emoji === undefined) throw new Error('Unknown')
```
Comment (AkiaCode) [inline src/test/cmds/addemoji.ts:16]: if (

**Review by AkiaCode — CHANGES_REQUESTED** [review]

File: src/structures/member.ts [hunk 31]
```diff
@@ -1,26 +1,28 @@
...
+import { Role } from './role.ts'
 import { User } from './user.ts'
 
 export class Member extends Base {
   id: string
   user: User
   nick?: string
-  roles: string[]
+  roleIDs: string[]
+  roles: Role[] = []
```
Comment (Helloyunho) [inline src/structures/member.ts:12]: I think this property can be removed

File: src/structures/member.ts [hunk 32]
```diff
@@ -32,7 +34,7 @@ export class Member extends Base {
   protected readFromData (data: MemberPayload): void {
     super.readFromData(data.user)
     this.nick = data.nick ?? this.nick
-    this.roles = data.roles ?? this.roles
+    this.roleIDs = data.roles ?? this.roles
```
Comment (Helloyunho) [inline src/structures/member.ts:37]: this.roleIDs wants `string[]` but this.roles is `Role[]`

**Review by Helloyunho — CHANGES_REQUESTED** [review]

File: src/structures/member.ts (hunk 31 above)
Comment (DjDeveloperr) [reply src/structures/member.ts:12]: Cannot be yet, it is used to retrieve role objects. In next PRs I'll add MemberRolesManager.

File: src/structures/member.ts (hunk 32 above)
Comment (DjDeveloperr) [reply src/structures/member.ts]: Fixed it

**Review by Helloyunho — APPROVED** [review]
I think it's good now

**Pull request merged**